onnxruntime/core/providers/cuda/tensor/tile_impl.cu

```diff
@@ -19,7 +21,12 @@ __global__ void _TileKernel(
   CALCULATE_ELEMENTWISE_INDEX_OR_EXIT(id, N);
   CUDA_LONG input_index = 0;
   CUDA_LONG output_index = id;
-  for (int dim = 0; dim < shape_rank; ++dim) {
+
+  #pragma unroll
+  for (int dim = 0; dim < MAX_DIMS; ++dim) {
```

Comment:
is it possible for `shape_rank` to be greater than `MAX_DIMS`?

Reply:
Good catch, will let those cases fallback. :)

Reply:
Fixed it.

Reply:
seems like there's a large performance difference between the old and unrolled implementations.
I don't fully understand why unrolling helps so much. just a thought, how would something like this do?
```c++
int dim = 0;
while (dim < shape_rank) {
  #pragma unroll
  for (int i = 0; i < MAX_DIMS; ++i, ++dim) {
    // ...
  }
}
```
if that performs ok, then perhaps we could avoid falling back to the significantly slower implementation. and there's only one kernel.

Reply:
I am not sure whether this would help ort not, because we introduced a while again. pragma is done in compiling stage.

Reply:
yeah, in the common case if there's one iteration, I see the additional cost of one extra comparison (if it were updated to `do {} while`). perhaps that's an optimistic view with my very incomplete knowledge :)

Reply:
sound reasonable, let me do some benchmark to see whether it works well. :)

Reply:
Unfortunately, using do-while the latency goes back to 560us for the test data set. So I am afraid we cannot use that trick.